driver-sync/src/test/functional/com/mongodb/client/MongoWriteConcernWithResponseExceptionTest.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client;
+
+import com.mongodb.Function;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.MongoWriteConcernException;
+import com.mongodb.ServerAddress;
+import com.mongodb.assertions.Assertions;
+import com.mongodb.event.CommandEvent;
+import com.mongodb.event.CommandFailedEvent;
+import com.mongodb.event.CommandListener;
+import com.mongodb.event.CommandSucceededEvent;
+import com.mongodb.internal.connection.MongoWriteConcernWithResponseException;
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonString;
+import org.bson.Document;
+import org.junit.Test;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.mongodb.ClusterFixture.isDiscoverableReplicaSet;
+import static com.mongodb.ClusterFixture.serverVersionAtLeast;
+import static com.mongodb.client.Fixture.getDefaultDatabaseName;
+import static com.mongodb.client.Fixture.getMongoClientSettingsBuilder;
+import static java.util.Collections.singletonList;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assume.assumeTrue;
+
+/**
+ * Tests in this class check that the internal {@link MongoWriteConcernWithResponseException} does not leak from our API.
+ */
+public final class MongoWriteConcernWithResponseExceptionTest {
```

Comment:
Should this follow the pattern that the other sync tests work by being overrideable / extendable and having a `createMongoClient` method?  

Would save adding another way of doing similar things.

Reply:
I followed the approach used in `RetryableWritesProseTest`. All we need here is to reuse code. Class inheritance allows doing so, but, much more importantly, it also allows to  have subtype polymorphism. To me, using inheritance solely for the purpose of code reuse (as is the case with tests) is weird, and I tend to avoid doing so.

The approach with inheritance was also not possible with `RetryableWritesProseTest` without refactoring the existing code structure: versions of `RetryableWritesProseTest` in sync/reactive modules already extend different versions of `DatabaseTestCase`.